[SYCLomatic][DOC] Add more manual fix examples in the document (#1230)
oneapi-src/SYCLomatic: docs/dev_guide/diagnostic_ref/dpct1065.rst

     auto tid = threadIdx.x + blockDim.x * blockIdx.x;
     process_a(data[tid].device_data);
     __syncthreads();
     process_b(data[tid].device_data);
   }

results in the following migrated SYCL\* code:

.. code-block:: cpp
   :linenos:

   struct Data_t {
     float *host_data;
     float *device_data;
   };
   
   void k(Data_t *data, const sycl::nd_item<3> &item_ct1) {
     auto tid = item_ct1.get_local_id(2) +
                item_ct1.get_local_range(2) * item_ct1.get_group(2);
     process_a(data[tid].device_data);
     /*
     DPCT1065:0: Consider replacing sycl::nd_item::barrier() with
     sycl::nd_item::barrier(sycl::access::fence_space::local_space) for better
     performance if there is no access to global memory.
     */
     item_ct1.barrier();
     process_b(data[tid].device_data);
   }

which is rewritten to:

.. code-block:: cpp
   :linenos:

   struct Data_t {
     float *host_data;
     float *device_data;
   };
   
   void k(Data_t *data, const sycl::nd_item<3> &item_ct1) {
     auto tid = item_ct1.get_local_id(2) +
                item_ct1.get_local_range(2) * item_ct1.get_group(2);
     process_a(data[tid].device_data);

>      item_ct1.barrier();

Reviewer: Please desc to give example on when can fix to high performance version: sycl::nd_item::barrier(sycl::access::fence_space::local_space).

     auto tid = threadIdx.x + blockDim.x * blockIdx.x;
     only_read_data(data[tid].device_data);
     __syncthreads();
     only_read_data(data[tid].device_data);
   }

results in the following migrated SYCL\* code:

.. code-block:: cpp
   :linenos:

   struct Data_t {
     float *host_data;
     float *device_data;
   };
   
   void k(Data_t *data, const sycl::nd_item<3> &item_ct1) {
     auto tid = item_ct1.get_local_id(2) +
                item_ct1.get_local_range(2) * item_ct1.get_group(2);
     only_read_data(data[tid].device_data);
     /*
     DPCT1065:0: Consider replacing sycl::nd_item::barrier() with
     sycl::nd_item::barrier(sycl::access::fence_space::local_space) for better
     performance if there is no access to global memory.
     */
     item_ct1.barrier();
     only_read_data(data[tid].device_data);
   }

which is rewritten to:

.. code-block:: cpp
   :linenos:

   struct Data_t {
     float *host_data;
     float *device_data;
   };
   
   void k(Data_t *data, const sycl::nd_item<3> &item_ct1) {
     auto tid = item_ct1.get_local_id(2) +
                item_ct1.get_local_range(2) * item_ct1.get_group(2);
     only_read_data(data[tid].device_data);

>      // global_local_space can be replaced with local_space if the access

Author: RAW, RAR, WAW

>      // We can use local_space to replace global_local_space if the access

Author: ```suggestion
     // global_local_space  can be replaced by local_space if the access

```

Author: done

     // of the global memory after the barrier does not depend on the access

>      // of the same global memory before the barrier.

Author: among threads in a WA

     performance if there is no access to global memory.
     */
     item_ct1.barrier();
     only_read_data(data[tid].device_data);
     /*
     DPCT1065:1: Consider replacing sycl::nd_item::barrier() with
     sycl::nd_item::barrier(sycl::access::fence_space::local_space) for better
     performance if there is no access to global memory.
     */
     item_ct1.barrier();
     only_write_data(data[tid].device_data);
     /*
     DPCT1065:2: Consider replacing sycl::nd_item::barrier() with
     sycl::nd_item::barrier(sycl::access::fence_space::local_space) for better
     performance if there is no access to global memory.
     */
     item_ct1.barrier();
     only_write_data(data[tid].device_data);
   }

which is rewritten to:

.. code-block:: cpp
   :linenos:

   struct Data_t {
     float *host_data;
     float *device_data;
   };
   
   void k(Data_t *data, const sycl::nd_item<3> &item_ct1) {
     auto tid = item_ct1.get_local_id(2) +
                item_ct1.get_local_range(2) * item_ct1.get_group(2);
     only_read_data(data[tid].device_data);
     item_ct1.barrier(sycl::access::fence_space::local_space);
     only_read_data(data[tid].device_data);

>      item_ct1.barrier();

Reviewer: add comment to say it's possible to use local_space, depend on the data access pattern

Author: done
